fix: Mitigate editor assets endpoint hook errors

dcalhoun opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fix CMM-937. 

Sometimes, plugins conditionally define functions based on `is_admin()`. For REST API endpoints, `is_admin()` is false. This can result in invocations of undefined functions when applying hooks—e.g., `enqueue_block_editor_assets`. 

As a specific example, [`wpforms-lite`](https://wordpress.org/plugins/wpforms-lite/) conditionally [registers `wpforms_admin_upgrade_link`](https://plugins.trac.wordpress.org/browser/wpforms-lite/trunk/src/WPForms.php?rev=3367762#L242) during `plugins_loaded`. Because `plugins_loaded` occurs before the REST API endpoint runs, we cannot influence `is_admin()`'s return before `wpforms-lite` evaluates whether to define its function. This results in a fatal error when applying `enqueue_block_editor_assets`. 

<details><summary>Example wpforms-lite fatal error</summary>
<p>

```
PHP Fatal error:  Uncaught Error: Call to undefined function WPForms\Admin\Education\wpforms_admin_upgrade_link() in /srv/htdocs/wp-content/plugins/wpforms-lite/src/Admin/Education/StringsTrait.php:178
Stack trace:
#0 /srv/htdocs/wp-content/plugins/wpforms-lite/src/Admin/Education/StringsTrait.php(97): WPForms\Integrations\Gutenberg\FormSelector->get_upgrade_strings('Pro', '%name%', 'gutenberg')
#1 /srv/htdocs/wp-content/plugins/wpforms-lite/src/Integrations/Gutenberg/FormSelector.php(404): WPForms\Integrations\Gutenberg\FormSelector->get_js_strings()
#2 /srv/htdocs/wp-content/plugins/wpforms-lite/src/Lite/Integrations/Gutenberg/FormSelector.php(80): WPForms\Integrations\Gutenberg\FormSelector->enqueue_block_editor_assets()
#3 /srv/htdocs/__wp__/wp-includes/class-wp-hook.php(324): WPForms\Lite\Integrations\Gutenberg\FormSelector->enqueue_block_editor_assets('')
#4 /srv/htdocs/__wp__/wp-includes/class-wp-hook.php(348): WP_Hook->apply_filters(NULL, Array)
#5 /srv/htdocs/__wp__/wp-includes/plugin.php(517): WP_Hook->do_action(Array)
#6 /srv/htdocs/wp-content/plugins/jetpack-dev/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-block-editor-assets.php(220): do_action('enqueue_block_e...')
#7 /srv/htdocs/__wp__/wp-includes/rest-api/class-wp-rest-server.php(1292): WPCOM_REST_API_V2_Endpoint_Block_Editor_Assets->get_items(Object(WP_REST_Request))
#8 /srv/htdocs/__wp__/wp-includes/rest-api/class-wp-rest-server.php(1125): WP_REST_Server->respond_to_request(Object(WP_REST_Request), '/wpcom/v2/edito...', Array, NULL)
#9 /srv/htdocs/__wp__/wp-includes/rest-api/class-wp-rest-server.php(439): WP_REST_Server->dispatch(Object(WP_REST_Request))
#10 /srv/htdocs/__wp__/wp-includes/rest-api.php(459): WP_REST_Server->serve_request('/wpcom/v2/edito...')
#11 /srv/htdocs/__wp__/wp-includes/class-wp-hook.php(324): rest_api_loaded(Object(WP))
#12 /srv/htdocs/__wp__/wp-includes/class-wp-hook.php(348): WP_Hook->apply_filters('', Array)
#13 /srv/htdocs/__wp__/wp-includes/plugin.php(565): WP_Hook->do_action(Array)
#14 /srv/htdocs/__wp__/wp-includes/class-wp.php(418): do_action_ref_array('parse_request', Array)
#15 /srv/htdocs/__wp__/wp-includes/class-wp.php(818): WP->parse_request('')
#16 /srv/htdocs/__wp__/wp-includes/functions.php(1342): WP->main('')
#17 /srv/htdocs/__wp__/wp-blog-header.php(16): wp()
#18 /srv/htdocs/__wp__/index.php(17): require('/srv/htdocs/__w...')
#19 {main}
  thrown in /srv/htdocs/wp-content/plugins/wpforms-lite/src/Admin/Education/StringsTrait.php on line 178
```

</p>
</details> 

We cannot catch the fatal errors with try-catch as they occur in a different context from the invocation of `do_action( 'enqueue_block_editor_assets' );`—WordPress invokes each hook itself, we do not. 

To workaround this, the changes remove hooks based on name and/or file paths. This workaround is not ideal, but appears the only option. Gutenberg core is [experimenting](https://github.com/WordPress/gutenberg/pull/73028) with REST API endpoints for editor assets as well; presumably that exploration will eventually encounter this issue and identify a core-provided solution. 

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
Strategically remove problematic hooks from plugins to avoid fatal errors. 

### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [x] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
N/A

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

1. Install and activate the [`wpforms-lite`](https://wordpress.org/plugins/wpforms-lite/) plugin for your site. 
2. Perform a request to the editor assets endpoint via the API Console, `curl`, etc. 
3. **Verify** a 200 response including the expected editor assets.

dcalhoun committed:
fix: Disable problematic wpforms-lite hooks

`is_admin()` specific logic invocations results in fatal errors. this
occurs because the REST API is not within WP Admin.

```
PHP Fatal error:  Uncaught Error: Call to undefined function WPForms\Admin\Education\wpforms_admin_upgrade_link() in /srv/htdocs/wp-content/plugins/wpforms-lite/src/Admin/Education/StringsTrait.php:178
Stack trace:
  thrown in /srv/htdocs/wp-content/plugins/wpforms-lite/src/Admin/Education/StringsTrait.php on line 178
```

dcalhoun committed:
test: Assert editor assets endpoint excludes problematic hooks

dcalhoun committed:
task: Catch potential Reflection errors

dcalhoun committed:
refactor: Hook removal leverages plugin variables for paths

Handle non-standard locations.

dcalhoun committed:
docs: Further document the necessity for hook removal

dcalhoun committed:
changelog

dcalhoun committed:
fix: Address phan error from test mock

```
FILE: projects/plugins/jetpack/tests/php/core-api/wpcom-endpoints/WPCOM_REST_API_V2_Endpoint_Block_Editor_Assets_Test.php
-------------------------------------------------------------------------------------------------------------------------------------------------------------
FOUND 4 ISSUES
-------------------------------------------------------------------------------------------------------------------------------------------------------------
1185 | UndefError PhanUndeclaredClassStaticProperty Reference to static property callback_executed from undeclared class \WPForms_Mock_Class
1188 | UndefError PhanUndeclaredClassMethod Call to method __construct from undeclared class \WPForms_Mock_Class
1191 | UndefError PhanUndeclaredClassInCallable Reference to undeclared class \WPForms_Mock_Class in callable \WPForms_Mock_Class::mock_callback
1204 | UndefError PhanUndeclaredClassStaticProperty Reference to static property callback_executed from undeclared class \WPForms_Mock_Class
-------------------------------------------------------------------------------------------------------------------------------------------------------------
```

dcalhoun committed:
task: Handle array-based callbacks

dcalhoun committed:
refactor: Improve test stability

dcalhoun committed:
test: Assert array callback patterns

dcalhoun commented on projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-block-editor-assets.php:
```diff
@@ -469,6 +472,133 @@ private function setup_block_editor_screen() {
+				} catch ( ReflectionException $e ) { // phpcs:ignore Generic.CodeAnalysis.EmptyStatement.DetectedCatch
+					// Failed to reflect on this object (e.g., internal class, anonymous class).
+					// Continue checking other callback types.
+				}
```
Addressed in https://github.com/Automattic/jetpack/pull/45924/commits/a8d5b81ef8ac58b4a95750ff0963d8d7200a4804.

dcalhoun commented on projects/plugins/jetpack/tests/php/core-api/wpcom-endpoints/WPCOM_REST_API_V2_Endpoint_Block_Editor_Assets_Test.php:
```diff
@@ -1142,4 +1142,76 @@ function () {
+		if ( $created_plugin_file ) {
+			unlink( $plugin_file );
+		}
+		if ( $created_plugin_dir ) {
```
Addressed in https://github.com/Automattic/jetpack/pull/45924/commits/6877a88589b5092bd112d43aba40a8b16542532e.

enejb commented on projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-block-editor-assets.php:
```diff
@@ -469,6 +472,149 @@ private function setup_block_editor_screen() {
+				'enqueue_block_editor_assets',
+			),
+		);
+
```
It might be good to just check if the plugin is active before doing all this work. 

You can do this using the `is_plugin_active()` function https://developer.wordpress.org/reference/functions/is_plugin_active/

dcalhoun replied:
Addressed this in 0284583885324e73cfab793dede58d46cb4b24ec.

enejb commented on projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-block-editor-assets.php:
```diff
@@ -469,6 +472,149 @@ private function setup_block_editor_screen() {
+				}
+
+				// Iterate through all priorities for this hook
+				foreach ( $wp_filter[ $hook_name ]->callbacks as $priority => $callbacks ) {
```
There is a lot of loops here. 

I wonder if we can cache some of this info into an array. 

Something like

Hook_callbacks_method_to_path array. Where we figure out the path of each of the callback once. 

And then we do another loop where we do the remove_action based on the plugin.

dcalhoun replied:
Good point. I addressed this in 0284583885324e73cfab793dede58d46cb4b24ec, and opted to simplify the implementation to target the one problematic hook usage: `wpforms-lite` with `enqueue_block_editor_assets`, while retaining the ability to add additional plugins. 

Let me know if I misunderstood or overlooked anything.

enejb commented on projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-block-editor-assets.php:
```diff
@@ -469,6 +472,149 @@ private function setup_block_editor_screen() {
+				'enqueue_block_editor_assets',
+			),
+		);
+
```
Usually how I would approach this would just be offending callback with a direct remote_action( '' ) call. but this doesn't seem possible with the way the `wpforms-lite` structures its classes as far as I can tell.

dcalhoun replied:
Agreed. Because of how `wpforms-lite` registers its callback, we need to detect it via reflection and file path.

enejb commented on projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-block-editor-assets.php:
```diff
@@ -469,6 +472,149 @@ private function setup_block_editor_screen() {
+			}
+		}
+
+		// Handle static method callbacks 'ClassName::method_name'
```
These case and the function case might not be needed since we could  directly remove the offending code directly if the plugin was using such methods.

dcalhoun replied:
Good point. Addressed this in 0284583885324e73cfab793dede58d46cb4b24ec.

enejb commented on projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-block-editor-assets.php:
```diff
@@ -469,6 +472,149 @@ private function setup_block_editor_screen() {
+				}
+			}
+
+			// Handle static method array callbacks ['ClassName', 'method_name']
```
This also could be removed since we could just directly call 
something like 
```
remove_action( 'hook', ['ClassName', 'method_name'] )
```

dcalhoun replied:
Good point. Addressed this in 0284583885324e73cfab793dede58d46cb4b24ec.

dcalhoun committed:
refactor: Simplify problematic hook removal

Solely target `wpforms-lite`: only `enqueue_block_editor_assets` and
object method callbacks.

Avoid performance costs of nest loops; remove unused complexity.

enejb reviewed (approved):
Nice work. This fixed the fatal error for me with the wpforms-lite plugin.
